### PR TITLE
Rewriting tank dispensers.

### PR DESCRIPTION
**Pull request by MistakeNot4892**
## Description of changes
Rewrites some crufty old dispenser code and makes them craftable.

## Why and what will this PR improve
Better code, more crafting options.

## Authorship
Myself.

## Changelog
:cl:
add: Tank dispensers are now craftable.
/:cl:

### CODE REVIEW

File: code/game/objects/structures/tank_dispenser.dm [hunk 1]
```diff
@@ -1,117 +1,135 @@
...
+	return ..()
+
+/obj/structure/tank_dispenser/Topic(href, href_list)
+
+	. = ..()
+	if(. || !usr || usr.incapacitated() || usr.restrained())
+		return
+
+	if(!Adjacent(usr))
 		close_browser(usr, "window=dispenser")
```
Comment (afterthought2) [inline code/game/objects/structures/tank_dispenser.dm]: Check the topic state, use OnTopic, at least something along those lines.

File: code/game/objects/structures/tank_dispenser.dm [hunk 2]
```diff
@@ -1,117 +1,135 @@
...
+		var/obj/item/tank/H = hydrogen_tanks[1]
+		LAZYREMOVE(hydrogen_tanks, H)
+		H.dropInto(loc)
+		to_chat(usr, SPAN_NOTICE("You take \the [H] out of \the [src]."))
+		. = TOPIC_REFRESH
+
+	if(. == TOPIC_REFRESH)
+		usr.set_machine(src)
+		update_icon()
+		add_fingerprint(usr)
```
Comment (afterthought2) [inline code/game/objects/structures/tank_dispenser.dm]: Don't add fingerprints. Setting machine here is a giant hack but apparently I'm in a forgiving mood when it comes to the machine var at the moment.

File: code/game/objects/structures/tank_dispenser.dm [hunk 3]
```diff
@@ -1,117 +1,135 @@
...
+		for(var/i in 1 to spawn_oxy)
+			oxygen_tanks += new /obj/item/tank/oxygen(src)
+	else
+		oxygen_tanks = null
+
+	if(isnum(hydrogen_tanks) && hydrogen_tanks > 0)
+		var/spawn_hyd = hydrogen_tanks
+		hydrogen_tanks = list()
+		for(var/i in 1 to spawn_hyd)
+			hydrogen_tanks += new /obj/item/tank/hydrogen(src)
```
Comment (afterthought2) [inline code/game/objects/structures/tank_dispenser.dm]: I wouldn't complain about this if it was oldcode, but you actually changed it to work like this: do we really need to spawn 20 tanks whenever one of these is mapped in or created?

File: code/game/objects/structures/tank_dispenser.dm [hunk 4]
```diff
@@ -1,117 +1,135 @@
...
+	switch(hydrocount)
 		if(1 to 4)
-			add_overlay("hydrogen-[hydrogentanks]")
+			add_overlay("hydrogen-[hydrocount]")
 		if(5 to INFINITY)
 			add_overlay("hydrogen-5")
 
-/obj/structure/dispenser/attack_ai(mob/living/silicon/ai/user)
+/obj/structure/tank_dispenser/attack_ai(mob/living/silicon/ai/user)
 	if(user.Adjacent(src))
```
Comment (afterthought2) [inline code/game/objects/structures/tank_dispenser.dm]: This check is unreasonable, use topic states for sane interactivity checks.

**Review by afterthought2 — COMMENTED** [review]
- It should really delete all of its tanks in Destroy. 
- Please double-check that in the physical destroy proc it doesn't dump its tanks while retaining refs to them.
- If the physical destroy handling involves dumping tanks, suggest reconsidering to destroy them instead for performance reasons.
- If it's buildable it needs to be deconstructable; maybe it is due to parent code but can't tell.

File: code/game/objects/structures/tank_dispenser.dm (hunk 3 above)
Comment (MistakeNot4892) [reply code/game/objects/structures/tank_dispenser.dm]: I would argue yes, if we're intending down the track to do physics sim etc. We spawn the contents of crates and lockers, we spawn organs, we populate with chems. I don't think tanks are that much more of an ask.

Comment (MistakeNot4892) [reply code/game/objects/structures/tank_dispenser.dm]: I was thinking of dropping the tank count though, I've never seen a tank dispenser get emptied in a round.

File: code/game/objects/structures/tank_dispenser.dm (hunk 4 above)
Comment (MistakeNot4892) [reply code/game/objects/structures/tank_dispenser.dm]: I might just completely rewrite the interaction and topic handling to be honest.

File: code/game/objects/structures/tank_dispenser.dm (hunk 1 above)
Comment (MistakeNot4892) [reply code/game/objects/structures/tank_dispenser.dm]: Done.

File: code/game/objects/structures/tank_dispenser.dm (hunk 2 above)
Comment (MistakeNot4892) [reply code/game/objects/structures/tank_dispenser.dm]: Removed.

File: code/game/objects/structures/tank_dispenser.dm (hunk 3 above)
Comment (MistakeNot4892) [reply code/game/objects/structures/tank_dispenser.dm]: Tank count has been dropped to six of each. Willing to discuss this further.

File: code/game/objects/structures/tank_dispenser.dm (hunk 4 above)
Comment (MistakeNot4892) [reply code/game/objects/structures/tank_dispenser.dm]: Done.

**Review by afterthought2 — APPROVED** [review]

**Pull request merged**